br/pkg/utils/retry.go

```diff
@@ -32,6 +36,130 @@ var retryableServerError = []string{
 	"end of file before message length reached",
 }
 
+type ErrorResult struct {
+	Strategy ErrorStrategy
+	Reason   string
+}
+
+type ErrorStrategy int
+
+const (
+	// This type can be retry but consume the backoffer attempts.
+	Retry ErrorStrategy = iota
+	// This type means un-recover error and the whole progress should exits
+	// for example:
+	// 1. permission not valid.
+	// 2. data has not found.
+	// 3. retry too many times
+	GiveUp
+)
+
+type ErrorContext struct {
+	mu sync.Mutex
+	// encounter times for one context on a store
+	// we may use this value to determine the retry policy
+	encounterTimes map[uint64]int
+	// unknown error retry limitation.
+	// encouter many times error makes Retry to GiveUp.
+	encounterTimesLimitation int
+	// whether in backup or restore
+	scenario string
+}
+
+func NewErrorContext(scenario string, limitation int) *ErrorContext {
+	return &ErrorContext{
+		scenario:                 scenario,
+		encounterTimes:           make(map[uint64]int),
+		encounterTimesLimitation: limitation,
+	}
+}
+
+func NewDefaultContext() *ErrorContext {
+	return &ErrorContext{
+		encounterTimes:           make(map[uint64]int),
+		encounterTimesLimitation: 1,
+	}
+}
+
+func (ec *ErrorContext) HandleError(err *backuppb.Error, uuid uint64, canIgnore bool) ErrorResult {
+	if len(err.Msg) != 0 {
+		return ec.HandleErrorMsg(err.Msg, uuid)
+	}
+	return ec.HandleErrorPb(err, uuid, canIgnore)
+}
+
+func (ec *ErrorContext) HandleErrorMsg(msg string, uuid uint64) ErrorResult {
+	// UNSAFE! TODO: use meaningful error code instead of unstructured message to find failed to write error.
+	logger := log.L().With(zap.String("scenario", ec.scenario))
+	if messageIsNotFoundStorageError(msg) {
+		reason := fmt.Sprintf("File or directory not found on TiKV Node (store id: %v). "+
+			"work around:please ensure br and tikv nodes share a same storage and the user of br and tikv has same uid.",
+			uuid)
```

Comment:
Why store id is a uuid?🤔

Reply:
because during restore. we cannot get store id